### PR TITLE
Add a periodic task (cron) system

### LINKED ISSUE #85: Add a very simple cron mechanism
I'm being told the project would be easier to use if we provided a simple cron-like process. It might be interesting to look.

@EliotBerriot, you mentionned you were maybe interested to try a contribution ?

### PR DESCRIPTION
**Pull request by ewjoachim**
Closes #85 
Closes #92 

### Successful PR Checklist:

- [x] Works
- [x] Tests
  - [ ] (not applicable?)
- [x] Documentation
  - [ ] (not applicable?)
- [x] Had a good time contributing?

### CODE REVIEW

**Comment by ewjoachim** [comment]
Lacking doc & tests, but it works. I'd like to do a first round of review to validate the API before I invest too much. Happy to to a guided tour with anyone interested.

**Comment by ewjoachim** [comment]
Lacking tests and then we're good !

**Comment by ewjoachim** [comment]
Just missing acceptance tests \o/

File: docs/howto/locks.rst [hunk 1]
```diff
@@ -1,44 +1,30 @@
...
 
-Or simply::
+.. warning::
 
-    job_description = write_alphabet.configure(lock="/tmp/alphabet.txt")
-    job_description.defer(letter="a")
-    job_description.defer(letter="b")
-    job_description.defer(letter="c")
-    job_description.defer(letter="d")
+    If a task is deferred with a lock and it has a ``scheduled_at`` arguments, then
```
Comment (CorBott) [inline docs/howto/locks.rst:25]:
```suggestion
    If a task with a "scheduled at" argument is deferred with a lock, then 
```

File: docs/howto/locks.rst [hunk 2]
```diff
@@ -1,44 +1,30 @@
...
+    following tasks will still not run until after the task has been processed, which
```
Comment (CorBott) [inline docs/howto/locks.rst:26]:
```suggestion
    following tasks can only run after this one is processed, which 
```

File: docs/howto/locks.rst [hunk 3]
```diff
@@ -1,44 +1,30 @@
...
+    may be arbitrary far in the future.
```
Comment (CorBott) [inline docs/howto/locks.rst:27]:
```suggestion
    may be in a long time.
```

File: docs/howto/locks.rst [hunk 4]
```diff
@@ -1,44 +1,30 @@
...
 
-Both ways, we're assured of getting ``abcd`` in our file.
+    Similarily, if the oldest task of a lock is in a queue that no worker consumes, the
+    other tasks will be blocked.
```
Comment (CorBott) [inline docs/howto/locks.rst:30]:
```suggestion
    other tasks are blocked.
```

File: procrastinate/app.py [hunk 5]
```diff
@@ -77,15 +78,29 @@ def __init__(
...
             All the values passed here will override the default values sent when
             launching a worker. See `App.run_worker` for details.
+        periodic_defaults :
+            Parameters for fine tuning the periodic tasks deferrer. Available
+            parameters are:
+
+            - ``max_delay``: ``float``, in seconds, controls how long after the planned
+              launch of a periodic task a deferrer can launch the task. Thanks to this
+              parameter, when deploying a new periodic task, it's usually not deferred
+              until its next scheduled time. Default to 10 minutes.
```
Comment (CorBott) [inline procrastinate/app.py:88]:
```suggestion
              until its next scheduled time. Defaults to 10 minutes.
```

File: procrastinate/periodic.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+import time
+from typing import Dict, Iterable, List, Optional, Tuple
+
+import attr
+import croniter
+
+from procrastinate import store, tasks
+
+# The maximum delay after which tasks will be considered as
+# too late, and ignored.
```
Comment (CorBott) [inline procrastinate/periodic.py:13]:
```suggestion
# outdated, and ignored.
```

File: procrastinate/periodic.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    def __init__(self, job_store: store.JobStore, max_delay: float = MAX_DELAY):
+        self.periodic_tasks: List[PeriodicTask] = []
+        self.job_store = job_store
+        # {task_name: defer_timestamp}
+        self.last_defers: Dict[str, int] = {}
+        self.max_delay = max_delay
+
+    def periodic_decorator(self, cron: str):
+        """
+        Decorator over a task definition that will reegister that task for periodic
```
Comment (CorBott) [inline procrastinate/periodic.py:44]:
```suggestion
        Decorator over a task definition that registers that task for periodic
```

File: procrastinate/periodic.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        launch. This decorator should not be use directly, ``@app.periodic()`` is meant
```
Comment (CorBott) [inline procrastinate/periodic.py:45]:
```suggestion
        launch. This decorator should not be used directly, ``@app.periodic()`` is meant
```

File: procrastinate/periodic.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+        def wrapper(task: tasks.Task):
+            self.register_task(task=task, cron=cron)
+            return task
+
+        return wrapper
+
+    async def worker(self) -> None:
+        """
+        High level command for the periodic deferrer. Launches the loop.
```
Comment (CorBott) [inline procrastinate/periodic.py:57]:
```suggestion
        High-level command for the periodic deferrer. Launches the loop.
```

File: procrastinate/utils.py [hunk 10]
```diff
@@ -201,3 +202,34 @@ def _get_module_name(obj: Any) -> str:
 def get_full_path(obj: Any) -> str:
 
     return f"{_get_module_name(obj)}.{obj.__name__}"
+
+
+@contextlib.contextmanager
+def task_context(awaitable: Awaitable, name: str):
+    """
+    Take an awaitable, returns a context manager.
```
Comment (CorBott) [inline procrastinate/utils.py:210]:
```suggestion
    Take an awaitable, return a context manager.
```

**Comment by ewjoachim** [comment]
Wow thanks a lot @CorBott for all the wording reviews !

And thanks @elemoine for the code review !

File: tests/unit/test_utils.py [hunk 11]
```diff
@@ -229,3 +230,53 @@ def test_get_full_path():
...
+        done.append(True)
+        await asyncio.sleep(10)
+
+    caplog.set_level("DEBUG")
+
+    with utils.task_context(foo(), name="foo") as task:
+        await asyncio.sleep(0)
+
+    assert done == [True]
+    # Give the task a cycled to update
```
Comment (ewjoachim) [inline tests/unit/test_utils.py:262]:
```suggestion
    # Give the task a cycle to update
```

**Pull request merged**